lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -935,11 +935,33 @@ static Value getCommonOperand(Op op) {
 template <typename Op>
 static bool canonicalizeIdempotentInputs(Op op, PatternRewriter &rewriter) {
   auto inputs = op.getInputs();
-  llvm::SmallSetVector<Value, 8> uniqueInputs;
 
-  for (const auto input : inputs)
-    uniqueInputs.insert(input);
+  llvm::SmallSetVector<Value, 8> alreadyCovered;
+  llvm::SmallSetVector<Value, 8> checked;
+  checked.insert(op);
 
+  llvm::SmallVector<Value, 8> worklist;
+  for (auto input : inputs) {
+    if (input != op)
+      worklist.push_back(input);
+  }
+
+  while (!worklist.empty()) {
+    auto element = worklist.pop_back_val();
+    checked.insert(element);
+
+    if (auto idempotentOp = element.getDefiningOp<Op>()) {
+      for (auto input : idempotentOp.getInputs()) {
+        alreadyCovered.insert(input);
+
+        if (!checked.contains(input))
+          worklist.push_back(input);
+      }
+    }
+  }
```

Comment:
The `insert` method on sets and maps returns a boolean indicating whether the element was inserted (true), or whether there was already an element in the set and nothing was added (false):
```suggestion
  while (!worklist.empty()) {
    auto element = worklist.pop_back_val();

    if (auto idempotentOp = element.getDefiningOp<Op>()) {
      for (auto input : idempotentOp.getInputs()) {
        alreadyCovered.insert(input);

        if (checked.insert(input))
          worklist.push_back(input);
      }
    }
  }
```